### PR TITLE
Cleanup AWS EC2 eventual consistency warnings

### PR DESCRIPTION
**Pull request by hakman**
When a cluster is created, there are warnings such as:
```
W0727 09:19:16.521016    5204 executor.go:128] error running task "AutoscalingGroup/nodes-ap-northeast-2c.e2e-8e9eb293c0-ff1eb.test-cncf-aws.k8s.io" (9m58s remaining to succeed): error creating AutoscalingGroup: ValidationError: You must use a valid fully-formed launch template. Value (nodes.e2e-8e9eb293c0-ff1eb.test-cncf-aws.k8s.io) for parameter iamInstanceProfile.name is invalid. Invalid IAM Instance Profile name
	status code: 400, request id: 56bfbee6-2dad-4c52-902e-5a8ddf9f5734
```

This happens because of AWS EC2 eventual consistency issues:
https://docs.aws.amazon.com/AWSEC2/latest/APIReference/query-api-troubleshooting.html#eventual-consistency

### CODE REVIEW

**Comment by johngmyers** [comment]
Good question whether we prefer the individual tasks to sleep or letting the task scheduler retry after a delay.

**Comment by hakman** [comment]
> Good question whether we prefer the individual tasks to sleep or letting the task scheduler retry after a delay.

The task already retries until it succeeds, but the problem is the useless warnings. Unless there is a way to tell the task just to retry and not print a warning, seems best to keep this logic in each individual task.

**Comment by johngmyers** [comment]
We could create a new type which, if embedded in an `error`, causes `fi.RunTasks()` to print at info level instead of warning.

File: upup/pkg/fi/cloudup/awstasks/launchtemplate_target_api.go [hunk 1]
```diff
@@ -142,31 +141,8 @@ func (t *LaunchTemplate) RenderAWS(c *awsup.AWSAPITarget, a, ep, changes *Launch
...
-				klog.V(4).Infof("ErrorCode=%q, Message=%q", awsup.AWSErrorCode(err), awsup.AWSErrorMessage(err))
-			}
-		}
-
-		return err
-	}()
-	if err != nil {
-		return fmt.Errorf("failed to create aws launch template: %s", err)
+	if _, err = c.Cloud.EC2().CreateLaunchTemplate(input); err != nil {
+		return fmt.Errorf("error creating LaunchTemplate: %v", err)
```
Comment (hakman) [inline upup/pkg/fi/cloudup/awstasks/launchtemplate_target_api.go:145]: I never seen these warnings in my tests. Would like to see if this workaround is still needed.

File: upup/pkg/fi/cloudup/awstasks/launchtemplate_target_api.go [hunk 2]
```diff
@@ -142,31 +141,8 @@ func (t *LaunchTemplate) RenderAWS(c *awsup.AWSAPITarget, a, ep, changes *Launch
...
 	// @step: attempt to create the launch template
-	err = func() error {
-		for attempt := 0; attempt < 10; attempt++ {
-			if _, err = c.Cloud.EC2().CreateLaunchTemplate(input); err == nil {
-				return nil
-			}
-
-			if awsup.AWSErrorCode(err) == "ValidationError" {
-				message := awsup.AWSErrorMessage(err)
-				if strings.Contains(message, "not authorized") || strings.Contains(message, "Invalid IamInstance") {
```
Comment (hakman) [inline upup/pkg/fi/cloudup/awstasks/launchtemplate_target_api.go:153]: I never seen these warnings in my tests. Would like to see if the handling for these errors is still needed.

**Comment by hakman** [comment]
> We could create a new type which, if embedded in an `error`, causes `fi.RunTasks()` to print at info level instead of warning.

I think it should be better now.

File: upup/pkg/fi/cloudup/awstasks/autoscalinggroup.go [hunk 3]
```diff
@@ -299,7 +299,13 @@ func (v *AutoscalingGroup) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *Autos
 
 		// @step: attempt to create the autoscaling group for us
 		if _, err := t.Cloud.Autoscaling().CreateAutoScalingGroup(request); err != nil {
-			return fmt.Errorf("error creating AutoscalingGroup: %v", err)
+			code := awsup.AWSErrorCode(err)
+			message := awsup.AWSErrorMessage(err)
+			if code == "ValidationError" && strings.Contains(message, "Invalid IAM Instance Profile name") {
+				klog.V(4).Infof("error creating AutoscalingGroup: %s", message)
+				return fi.NewTryAgainLaterError("Waiting for the IAM Instance Profile to be propagated")
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/awstasks/autoscalinggroup.go:306]:
```suggestion
				return fi.NewTryAgainLaterError("waiting for the IAM Instance Profile to be propagated")
```

File: upup/pkg/fi/cloudup/awstasks/launchconfiguration.go [hunk 4]
```diff
@@ -353,32 +351,14 @@ func (_ *LaunchConfiguration) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *La
...
-		klog.V(8).Infof("AWS CreateLaunchConfiguration %s", aws.StringValue(request.LaunchConfigurationName))
-		_, err = t.Cloud.Autoscaling().CreateLaunchConfiguration(request)
-		if err == nil {
-			break
+	if _, err = t.Cloud.Autoscaling().CreateLaunchConfiguration(request); err != nil {
+		code := awsup.AWSErrorCode(err)
+		message := awsup.AWSErrorMessage(err)
+		if code == "ValidationError" && strings.Contains(message, "Invalid IamInstanceProfile") {
+			klog.V(4).Infof("error creating LaunchConfiguration: %s", message)
+			return fi.NewTryAgainLaterError("Waiting for the IAM Instance Profile to be propagated")
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/awstasks/launchconfiguration.go:359]:
```suggestion
			return fi.NewTryAgainLaterError("waiting for the IAM Instance Profile to be propagated")
```

File: upup/pkg/fi/cloudup/awstasks/route.go [hunk 5]
```diff
@@ -175,7 +171,13 @@ func (_ *Route) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *Route) error {
 
 		response, err := t.Cloud.EC2().CreateRoute(request)
 		if err != nil {
-			return fmt.Errorf("error creating Route: %v", err)
+			code := awsup.AWSErrorCode(err)
+			message := awsup.AWSErrorMessage(err)
+			if code == "InvalidNatGatewayID.NotFound" {
+				klog.V(4).Infof("error creating Route: %s", message)
+				return fi.NewTryAgainLaterError("Waiting for the NAT Gateway to be created")
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/awstasks/route.go:178]:
```suggestion
				return fi.NewTryAgainLaterError("waiting for the NAT Gateway to be created")
```

File: upup/pkg/fi/cloudup/awstasks/route.go [hunk 6]
```diff
@@ -204,9 +202,14 @@ func (_ *Route) RenderAWS(t *awsup.AWSAPITarget, a, e, changes *Route) error {
...
 
-		_, err := t.Cloud.EC2().ReplaceRoute(request)
-		if err != nil {
-			return fmt.Errorf("error updating Route: %v", err)
+		if _, err := t.Cloud.EC2().ReplaceRoute(request); err != nil {
+			code := awsup.AWSErrorCode(err)
+			message := awsup.AWSErrorMessage(err)
+			if code == "InvalidNatGatewayID.NotFound" {
+				klog.V(4).Infof("error creating Route: %s", message)
+				return fi.NewTryAgainLaterError("Waiting for the NAT Gateway to be created")
```
Comment (johngmyers) [inline upup/pkg/fi/cloudup/awstasks/route.go:210]:
```suggestion
				return fi.NewTryAgainLaterError("waiting for the NAT Gateway to be created")
```

**Pull request merged**